### PR TITLE
perf(speculative): add fused Triton soft cross-entropy kernel for EAGLE-3

### PR DESCRIPTION
**Pull request by khazic**
# What does this PR do ?

Add a fused Triton soft-target cross-entropy kernel for EAGLE-3 / P-EAGLE draft training. The primary motivation is memory: the PyTorch `F.log_softmax` + elementwise multiply + sum path in `masked_soft_cross_entropy` materializes a full `[B, S, vocab]` float32 log-probs tensor, which is the dominant peak-memory term in draft training. The fused kernel computes the loss with a single-pass online softmax and never materializes that tensor, cutting the loss peak memory to roughly the inputs alone.

Measured savings scale with sequence length, e.g. ~12.6 GB saved at `[1, 32768, 32000]` (4x less peak memory for this op), with a speedup on top (memory-bound, so it varies by GPU bandwidth). See the benchmark comments below for A6000 and A100 numbers.

## Impact on P-EAGLE

This should let P-EAGLE train at sequence length 4096 without OOM, and without needing to enable sequence partitioning to fit. Sequence partitioning stays available for longer sequences, but it should no longer be required just to make 4096 fit.

# Changelog

- Add `nemo_automodel/components/loss/triton/soft_cross_entropy.py`: `fused_soft_cross_entropy` plus `SoftCrossEntropyFunction` (forward computes loss with a single-pass online softmax and saves `lse`; backward reuses `lse` to compute `softmax(x) - p` in one pass; masked rows early-return).
- `masked_soft_cross_entropy` (`soft_ce.py`) dispatches to the fused kernel when Triton is available and the input is on CUDA, and otherwise keeps the existing PyTorch path.
- Reuse `MAX_FUSED_SIZE` and `element_mul_kernel` from the sibling `te_cross_entropy.py` instead of redefining them.
- Add CUDA parity tests (`tests/unit_tests/loss/test_soft_ce_triton.py`) covering forward loss, backward gradients, all-masked, single-valid-position, fp32 targets, and one-hot targets against the PyTorch reference.
- Add a standalone microbenchmark (`benchmark/ce_loss.py`).

# Note for reviewers

`masked_soft_cross_entropy` selects the fused kernel transparently via `if HAVE_TRITON and logits.is_cuda`, rather than as an explicitly-constructed loss object the way `FusedLinearCrossEntropy` / `TEParallelCrossEntropy` are selected. This keeps it a drop-in for existing callers, but differs from the explicit-selection convention used elsewhere. Happy to switch to explicit selection (threaded through the EAGLE recipe) if preferred.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?

# Additional Information

- Numerical parity with the PyTorch path is covered by the new tests (the kernel keeps fp32 accumulation).
- The speedup is memory-bound, so the relative speedup is lower on higher-bandwidth GPUs while the memory savings stay the same across GPUs.

### CODE REVIEW

**Comment by khazic** [comment]
## Benchmark results (NVIDIA RTX A6000, bf16)

Fused Triton kernel vs the PyTorch `F.log_softmax` baseline. `Loss diff` is the absolute loss difference and `Grad cos` is the cosine similarity of the gradients against the PyTorch path (both confirm numerical parity).

### Shape benchmark (mask_ratio=0.8, dtype=bf16)

| Shape | PyTorch | Triton | Speedup | PT Mem | Tri Mem | Saved | Loss diff | Grad cos |
|---|---|---|---|---|---|---|---|---|
| [1, 128, 1024] | 0.32ms | 0.36ms | 0.9x | 0.00GB | 0.00GB | 0.00GB | 0.00e+00 | 0.99999988 |
| [1, 256, 8192] | 0.64ms | 0.35ms | 1.8x | 0.03GB | 0.01GB | 0.03GB | 0.00e+00 | 0.99999994 |
| [1, 1024, 32000] | 3.54ms | 0.63ms | 5.6x | 0.53GB | 0.13GB | 0.40GB | 0.00e+00 | 1.00000000 |
| [1, 2048, 32000] | 7.03ms | 1.06ms | 6.6x | 1.05GB | 0.26GB | 0.79GB | 0.00e+00 | 1.00000000 |
| [1, 4096, 32000] | 14.01ms | 1.97ms | 7.1x | 2.10GB | 0.52GB | 1.57GB | 0.00e+00 | 0.99999994 |
| [1, 8192, 32000] | 27.96ms | 3.95ms | 7.1x | 4.19GB | 1.05GB | 3.15GB | 0.00e+00 | 1.00000000 |
| [1, 16384, 32000] | 55.88ms | 7.25ms | 7.7x | 8.39GB | 2.10GB | 6.29GB | 0.00e+00 | 1.00000012 |
| [1, 32768, 32000] | 116.74ms | 14.52ms | 8.0x | 16.78GB | 4.19GB | 12.58GB | 0.00e+00 | 1.00000000 |

At the largest EAGLE-3 shape ([1, 32768, 32000]) the kernel is 8.0x faster and saves 12.58 GB of peak memory.

### Mask ratio benchmark (shape=[1, 4096, 32000], dtype=bf16)

| Mask ratio | Valid | PyTorch | Triton | Speedup | Loss diff | Grad cos |
|---|---|---|---|---|---|---|
| 1.0 | 4096 | 14.66ms | 2.37ms | 6.2x | 0.00e+00 | 1.0000000000 |
| 0.8 | 3268 | 14.01ms | 1.96ms | 7.2x | 0.00e+00 | 1.0000002384 |
| 0.5 | 1993 | 14.01ms | 1.42ms | 9.9x | 9.54e-07 | 1.0000000000 |
| 0.2 | 854 | 14.01ms | 0.93ms | 15.1x | 0.00e+00 | 0.9999999404 |
| 0.05 | 236 | 14.01ms | 0.78ms | 18.0x | 0.00e+00 | 1.0000000000 |

The masked-row early-return pays off as sparsity grows: at mask_ratio=0.05 the kernel reaches 18.0x while the PyTorch path stays flat (it still computes every row).

@HuiyingLi would you be able to run the same benchmark (`nemo_automodel/components/loss/benchmark/ce_loss.py`) on a Blackwell (B-series) GPU? Want to confirm the speedup and memory savings hold there before we finalize.

**Comment by khazic** [comment]
## Benchmark results (NVIDIA A100-SXM4-80GB, bf16)

Same benchmark on an A100. The PyTorch baseline is faster here than on the A6000, so the relative speedups are a bit lower, while the memory savings are identical (same shapes).

### Shape benchmark (mask_ratio=0.8, dtype=bf16)

| Shape | PyTorch | Triton | Speedup | PT Mem | Tri Mem | Saved | Loss diff | Grad cos |
|---|---|---|---|---|---|---|---|---|
| [1, 128, 1024] | 0.63ms | 0.59ms | 1.1x | 0.00GB | 0.00GB | 0.00GB | 0.00e+00 | 0.99999976 |
| [1, 256, 8192] | 0.64ms | 0.56ms | 1.1x | 0.03GB | 0.01GB | 0.03GB | 0.00e+00 | 1.00000000 |
| [1, 1024, 32000] | 1.65ms | 0.58ms | 2.8x | 0.53GB | 0.13GB | 0.40GB | 0.00e+00 | 1.00000012 |
| [1, 2048, 32000] | 3.20ms | 0.63ms | 5.1x | 1.05GB | 0.26GB | 0.79GB | 0.00e+00 | 1.00000000 |
| [1, 4096, 32000] | 6.11ms | 1.03ms | 5.9x | 2.10GB | 0.52GB | 1.57GB | 0.00e+00 | 1.00000000 |
| [1, 8192, 32000] | 12.12ms | 2.00ms | 6.1x | 4.19GB | 1.05GB | 3.15GB | 0.00e+00 | 1.00000000 |
| [1, 16384, 32000] | 24.19ms | 3.92ms | 6.2x | 8.39GB | 2.10GB | 6.29GB | 0.00e+00 | 1.00000012 |
| [1, 32768, 32000] | 48.27ms | 7.78ms | 6.2x | 16.78GB | 4.19GB | 12.58GB | 0.00e+00 | 1.00000000 |

At the largest EAGLE-3 shape ([1, 32768, 32000]) the kernel is 6.2x faster and saves 12.58 GB of peak memory.

### Mask ratio benchmark (shape=[1, 4096, 32000], dtype=bf16)

| Mask ratio | Valid | PyTorch | Triton | Speedup | Loss diff | Grad cos |
|---|---|---|---|---|---|---|
| 1.0 | 4096 | 6.21ms | 1.19ms | 5.2x | 0.00e+00 | 1.0000000000 |
| 0.8 | 3246 | 6.21ms | 1.06ms | 5.8x | 0.00e+00 | 1.0000000000 |
| 0.5 | 2086 | 6.21ms | 0.90ms | 6.9x | 0.00e+00 | 1.0000000000 |
| 0.2 | 836 | 6.22ms | 0.72ms | 8.6x | 0.00e+00 | 1.0000000000 |
| 0.05 | 198 | 6.21ms | 0.62ms | 10.1x | 0.00e+00 | 0.9999998808 |

Numerical parity holds (loss diff at 0 and gradient cosine at ~1.0 across all shapes and mask ratios).

**Comment by HuiyingLi** [comment]
## Benchmark results (NVIDIA H100 80GB HBM3, bf16)

Ran the same benchmark on an H100:

```bash
CUDA_VISIBLE_DEVICES=0 PYTHONPATH=$PWD python nemo_automodel/components/loss/benchmark/ce_loss.py
```

GPU reported by the script: `NVIDIA H100 80GB HBM3 | 84.4GB free / 84.9GB total`

### Shape benchmark (mask_ratio=0.8, dtype=bf16)

| Shape | PyTorch | Triton | Speedup | PT Mem | Tri Mem | Saved | Loss diff | Grad cos |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| [1, 128, 1024] | 0.19ms | 0.19ms | 1.0x | 0.00GB | 0.00GB | 0.00GB | 0.00e+00 | 0.99999976 |
| [1, 256, 8192] | 0.18ms | 0.18ms | 1.0x | 0.03GB | 0.01GB | 0.03GB | 0.00e+00 | 1.00000012 |
| [1, 1024, 32000] | 0.94ms | 0.19ms | 5.0x | 0.53GB | 0.13GB | 0.40GB | 0.00e+00 | 1.00000012 |
| [1, 2048, 32000] | 1.82ms | 0.32ms | 5.6x | 1.05GB | 0.26GB | 0.79GB | 0.00e+00 | 1.00000000 |
| [1, 4096, 32000] | 3.59ms | 0.60ms | 5.9x | 2.10GB | 0.52GB | 1.57GB | 0.00e+00 | 1.00000012 |
| [1, 8192, 32000] | 7.11ms | 1.16ms | 6.1x | 4.19GB | 1.05GB | 3.15GB | 0.00e+00 | 1.00000024 |
| [1, 16384, 32000] | 14.15ms | 2.28ms | 6.2x | 8.39GB | 2.10GB | 6.29GB | 0.00e+00 | 1.00000012 |
| [1, 32768, 32000] | 28.27ms | 4.52ms | 6.3x | 16.78GB | 4.19GB | 12.58GB | 0.00e+00 | 0.99999994 |

At the largest EAGLE-3 shape (`[1, 32768, 32000]`), the fused Triton kernel is 6.3x faster and saves 12.58 GB of peak memory on H100.

### Mask ratio benchmark (shape=[1, 4096, 32000], dtype=bf16)

| Mask ratio | Valid | PyTorch | Triton | Speedup | Loss diff | Grad cos |
|---:|---:|---:|---:|---:|---:|---:|
| 1.0 | 4096 | 3.77ms | 0.68ms | 5.6x | 0.00e+00 | 1.0000001192 |
| 0.8 | 3286 | 3.58ms | 0.61ms | 5.9x | 0.00e+00 | 1.0000000000 |
| 0.5 | 2069 | 3.59ms | 0.50ms | 7.2x | 0.00e+00 | 1.0000000000 |
| 0.2 | 808 | 3.58ms | 0.39ms | 9.1x | 0.00e+00 | 1.0000000000 |
| 0.05 | 210 | 3.59ms | 0.34ms | 10.5x | 0.00e+00 | 1.0000000000 |

Numerical parity holds across the run (`Loss diff` is zero and gradient cosine is approximately 1.0 for all cases).

File: nemo_automodel/components/loss/benchmark/soft_ce_loss.py
Comment (HuiyingLi) [inline nemo_automodel/components/loss/benchmark/soft_ce_loss.py:1]: what about changing the file name to soft_ce_loss instead

Comment (khazic) [reply nemo_automodel/components/loss/benchmark/soft_ce_loss.py:1]: Good call, renamed to soft_ce_loss.py in 3bd4d859.

File: nemo_automodel/components/loss/triton/soft_cross_entropy.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        x_block = x_block.to(tl.float32)
+        p_block = tl.load(P_row_ptr + offsets, mask=valid, other=0.0).to(tl.float32)
+
+        grad_block = (tl.exp(x_block - lse) - p_block) * inv_count
+        tl.store(X_row_ptr + offsets, grad_block.to(out_dtype), mask=valid)
+
+
+def _select_num_warps(V: int) -> int:
+    if V <= 1024:
+        return 4
```
[the remark this answers is not in the document]
Comment (khazic) [reply nemo_automodel/components/loss/triton/soft_cross_entropy.py:147]: Fixed in 97190781. The forward now saves sum(p) per row (it already computed it) and the backward applies sum(p) * softmax(x) - p, so it's exact for arbitrary targets and consistent with the forward / PyTorch reference. Added an un-normalized-target forward+backward parity test.

**Review by HuiyingLi — APPROVED** [review]

**Comment by HuiyingLi** [comment]
## Benchmark results (NVIDIA GB200, bf16)

Ran the same benchmark on a GB200 GPU using final PR head `97190781`:

```bash
CUDA_VISIBLE_DEVICES=0 PYTHONPATH=$PWD python -u nemo_automodel/components/loss/benchmark/soft_ce_loss.py
```

GPU reported by the script: `NVIDIA GB200 | 197.2GB free / 197.9GB total`

### Shape benchmark (mask_ratio=0.8, dtype=bf16)

| Shape | PyTorch | Triton | Speedup | PT Mem | Tri Mem | Saved | Loss diff | Grad cos |
|---|---:|---:|---:|---:|---:|---:|---:|---:|
| [1, 128, 1024] | 0.44ms | 0.39ms | 1.1x | 0.00GB | 0.00GB | 0.00GB | 0.00e+00 | 0.99999994 |
| [1, 256, 8192] | 0.50ms | 0.42ms | 1.2x | 0.03GB | 0.01GB | 0.03GB | 9.54e-07 | 1.00000000 |
| [1, 1024, 32000] | 0.59ms | 0.37ms | 1.6x | 0.53GB | 0.13GB | 0.40GB | 0.00e+00 | 1.00000000 |
| [1, 2048, 32000] | 1.08ms | 0.38ms | 2.8x | 1.05GB | 0.26GB | 0.79GB | 1.91e-06 | 0.99999994 |
| [1, 4096, 32000] | 2.05ms | 0.39ms | 5.3x | 2.10GB | 0.52GB | 1.57GB | 0.00e+00 | 1.00000012 |
| [1, 8192, 32000] | 3.98ms | 0.69ms | 5.8x | 4.19GB | 1.05GB | 3.15GB | 0.00e+00 | 0.99999994 |
| [1, 16384, 32000] | 7.85ms | 1.32ms | 6.0x | 8.39GB | 2.10GB | 6.29GB | 0.00e+00 | 1.00000000 |
| [1, 32768, 32000] | 15.62ms | 2.59ms | 6.0x | 16.78GB | 4.19GB | 12.58GB | 0.00e+00 | 0.99999988 |

At the largest EAGLE-3 shape (`[1, 32768, 32000]`), the fused Triton kernel is 6.0x faster and saves 12.58 GB of peak memory on GB200.

### Mask ratio benchmark (shape=[1, 4096, 32000], dtype=bf16)

| Mask ratio | Valid | PyTorch | Triton | Speedup | Loss diff | Grad cos |
|---:|---:|---:|---:|---:|---:|---:|
| 1.0 | 4096 | 2.05ms | 0.42ms | 4.9x | 0.00e+00 | 1.0000000000 |
| 0.8 | 3256 | 2.05ms | 0.38ms | 5.3x | 9.54e-07 | 1.0000000000 |
| 0.5 | 2021 | 2.05ms | 0.39ms | 5.3x | 0.00e+00 | 1.0000000000 |
| 0.2 | 802 | 2.05ms | 0.37ms | 5.5x | 0.00e+00 | 0.9999999404 |
| 0.05 | 216 | 2.05ms | 0.38ms | 5.4x | 0.00e+00 | 1.0000000000 |

Numerical parity holds across the run (`Loss diff` is near zero and gradient cosine is approximately 1.0 for all cases).
